### PR TITLE
Add Synced Spins and Polarizations

### LINKED ISSUE #103: Synced Polarizations and Spins
It would be nice to add new indefinite spins and polarizations that sync polarizations or spins of different particles together. This would be helpful for example when considering multiple photons coming from a laser, where the polarizations are necessarily equal between the photons.
I suggest a type such as 
`SyncedPolarization{N} <: AbstractIndefinitePolarization`, where `N` would be an integer to differentiate between multiple possible synced polarizations in the same process. For a process like `kkke -> ke` and the incoming Photons all having `SyncedPolarization{1}` set (and the other particles `AllSpin`/`AllPol`), there would then be only 16 total polarizations possible instead of 64 if `AllPol` had to be used for all particles. When generating code this makes a large difference, and because of intermediate result reuse, it is also better than generating the process with `PolX` and `PolY` once and summing.

It's conceivable to have an even more powerful implementation of such a type, allowing specific combinations of Pols/Spins, but I'm not sure if there would be any actual application of this.

### PR DESCRIPTION
**Pull request by AntonReinhard**
Fixes #103

This adds the `SyncedPolarization{N}` and `SyncedSpin{N}` types proposed in #103 and reworks the `multiplicity` function to work on `AbstractProcessDefinition`s instead of raw spins or pols, since this previous notion of `multiplicity` does not extend to synced spins/pols.
The new implementation of `multiplicity` is faster than the `prod.(spins_pols)` that was used previously, because it can be compiled away when the spins and pols of the process are known at compile time.

This update breaks QEDprocesses and will require a tiny update there.

### CODE REVIEW

File: src/implementations/process/spin_pols.jl [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+"""
+    multiplicity(proc::AbstractProcessDefinition)
+
+Return the number of spin and polarization combinations represented by `proc` total. This depends on the specific [`AbstractSpinOrPolarization`](@ref)s returned by [`spin_pols`](@ref) for `proc`.
+For example, a default Compton process with four indefinite spins/polarizations has a multiplicity of 2^4 = 16. A Compton process with many incoming photons that have synced polarizations
+will still have a multiplicity of 16.
+
+!!! note Performance
+    As long as [`incoming_spin_pols`](@ref) and [`outgoing_spin_pols`](@ref) can be evaluated at compile time, this function is completely compiled away.
```
Comment (szabo137) [inline src/implementations/process/spin_pols.jl:101]: At some point, we should add such statements to the docs of the respective function as well.

File: src/interfaces/particles/spin_pol.jl [hunk 2]
```diff
@@ -179,13 +179,39 @@ const PolY = PolarizationY
...
+Two polarizations are considered synced when they have the same value for `N`. This means that
+the resulting multiplicity will be 2 total for all particles with the same `SyncedPolarization`.
 
+Having a single `SyncedPolarization{N}` in a process is legal. In this case, it behaves just
+like an [`AllPolarization`](@ref) would.
+
+See also: [`multiplicity`](@ref)
+"""
+struct SyncedPolarization{N} <: AbstractIndefinitePolarization
+    function SyncedPolarization{N}() where {N}
```
Comment (szabo137) [inline src/interfaces/particles/spin_pol.jl:194]: Is it really necessary to expose the type parameter in the constructor, wouldn't it be more convenient to have a standard call constructor like 

```julia 
function SyncedPolarization(N::Int)
    return new{N}()
end
```
Then the type assert is not necessary. If you need the constructor like it's implemented, the constructor above should conveniently added as well.

**Review by szabo137 — COMMENTED** [review]
I really like the idea and have just minor concerns. 

TL;DR:
Especially the possibility to only iterate over "valid" spin/pol combinations seems very powerful. This will come in handy, if we start implementing the helicity amplitudes: https://arxiv.org/abs/hep-ph/9805445

File: src/interfaces/particles/spin_pol.jl (hunk 2 above)
Comment (AntonReinhard) [reply src/interfaces/particles/spin_pol.jl:194]: You're right, that is much nicer and requires fewer brackets that I definitely didn't forget about ten times while writing this code

File: src/implementations/process/spin_pols.jl (hunk 1 above)
Comment (AntonReinhard) [reply src/implementations/process/spin_pols.jl:101]: I added small notes to the `AbstractProcessDefinition` interface

**Review by szabo137 — APPROVED** [review]
